extension/amqp/listener.go

```diff
@@ -0,0 +1,132 @@
+	"github.com/mailru/easyjson"
+	"github.com/streadway/amqp"
+)
+
+// Ensure Listener implements sql.Listener
+var _ sql.Listener = &Listener{}
+
+type (
+	// Consume returns a channel of amqp.Delivery's and a related closer or an error
+	Consume func() (io.Closer, <-chan amqp.Delivery, error)
+
+	// Listener consumes messages from an queue
+	Listener struct {
+		consume              Consume
+		minReconnectInterval time.Duration
+		maxReconnectInterval time.Duration
+		logger               goengine.Logger
+	}
+)
+
+// NewListener returns a new Listener
+func NewListener(
+	consume Consume,
+	minReconnectInterval time.Duration,
+	maxReconnectInterval time.Duration,
+	logger goengine.Logger,
+) (*Listener, error) {
+	switch {
+	case consume == nil:
+		return nil, goengine.InvalidArgumentError("consume")
+	}
+
+	if logger == nil {
+		logger = goengine.NopLogger
+	}
+
+	return &Listener{
+		consume:              consume,
+		minReconnectInterval: minReconnectInterval,
+		maxReconnectInterval: maxReconnectInterval,
+		logger:               logger,
+	}, nil
+}
+
+// Listen receives messages from a queue, transforms them into a sql.ProjectionNotification and calls the trigger
+func (l *Listener) Listen(ctx context.Context, trigger sql.ProjectionTrigger) error {
+	var nextReconnect time.Time
+	reconnectInterval := l.minReconnectInterval
+	for {
+		select {
+		case <-ctx.Done():
+			return context.Canceled
+		default:
+		}
+
+		conn, deliveries, err := l.consume()
+		if err != nil {
+			l.logger.Error("failed to start consuming amqp messages", func(entry goengine.LoggerEntry) {
+				entry.Error(err)
+				entry.String("reconnect_in", reconnectInterval.String())
+			})
+
+			time.Sleep(reconnectInterval)
+			reconnectInterval *= 2
+			if reconnectInterval > l.maxReconnectInterval {
+				reconnectInterval = l.maxReconnectInterval
+			}
+			continue
+		}
+		reconnectInterval = l.minReconnectInterval
+		nextReconnect = time.Now().Add(reconnectInterval)
+
+		l.consumeMessages(ctx, conn, deliveries, trigger)
+
+		select {
+		case <-ctx.Done():
+			return context.Canceled
+		default:
+			time.Sleep(time.Until(nextReconnect))
+		}
+	}
+}
+
+func (l *Listener) consumeMessages(ctx context.Context, conn io.Closer, deliveries <-chan amqp.Delivery, trigger sql.ProjectionTrigger) {
+	defer func() {
+		if conn == nil {
+			return
+		}
+
+		if err := conn.Close(); err != nil {
+			l.logger.Error("failed to close amqp connection", func(entry goengine.LoggerEntry) {
+				entry.Error(err)
+			})
+		}
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case msg, ok := <-deliveries:
+			if !ok {
+				return
+			}
+
+			notification := &sql.ProjectionNotification{}
+			if err := easyjson.Unmarshal(msg.Body, notification); err != nil {
+				l.logger.Error("failed to unmarshal delivery, dropping message", func(entry goengine.LoggerEntry) {
+					entry.Error(err)
+				})
+				continue
+			}
+
+			if err := trigger(ctx, notification); err != nil {
```

Comment:
Passing your process-level `ctx` through to each request, or even making it a child context is not a good idea. Closing the process context should initiate a clean and graceful shutdown, but if you’ve attached the request context to the process context (or they are the same), then when the process context is canceled, the request context will also cancel, which means the request handler should abort processing the delivery. So, then instead of gracefully shutting down, the outstanding handler could abort and not even finish doing what it was supposed to do.